### DISCUSSION TITLE
Custom projection

### DISCUSSION THREAD

**Discussion by CAPITAINMARVEL** [BeanieODM/beanie#710] [Question]
data: Optional[dict] = await ctx.bot.db.Players.find_one(
            filter={'player_id': ctx.author.id}, 
            projection={'_id': False, 'cd':f'$cd.{cd_name}'}
            )

this is what im trying to do
the problem with projection 

```py
class ProjectTest(BaseModel):
    def __init(self, cd_name: str):
        self.cd_name = cd_name
    class Settings:
        projection = {"cd": f"$cd.{self.cd_name}"}
        ```

its that i can't add a custom string for my need

**Comment by CAPITAINMARVEL** [comment]
maybe i didnt figure out how

**Reply by mrrobinhood5** [reply]
what does your Document model look like? this looks like you are trying to use pymongo to do the search, not Beanie

**Reply by CAPITAINMARVEL** [reply]
```python

class Cooldown(BaseModel):
    rep: datetime = utcnow()
    pay: datetime = utcnow()
    rob: datetime = utcnow()

class Player(Document):
    cd: Cooldown
```
basically im trying to do a projection to only return one of the field but the issue is i dont figure out to do it dynamically
like by adding an attribute to specify the field to return

```python
class ProjectionCd(BaseModel):
    cd: datetime
    class Settings:
        projection = {"cd": f"cd.{field_name}"}
```
i tried add a argument with the constructor but it doesnt let me

**Reply by mrrobinhood5** [reply]
Thats not really what projections are for. You dont even need the database functions to do that. An easy way to implement what you need is:
```py
    class Cooldown(BaseModel):
        rep: datetime = datetime.utcnow()
        pay: datetime = datetime.utcnow()
        rob: datetime = datetime.utcnow()

    class Player(Document):
        cooldown: Cooldown

        def cd(self, field) -> str:
            f = cooldown.model_dump().get(field)
            return f'cd.{f}'

        class Settings:
            name = 'CooldownTesting'
 ```
 then you would just call the cd() method with whatever field name you need from Cooldown

**Reply by CAPITAINMARVEL** [reply]
Thanks for the answer but i think I wasn’t clear
def cd(self, field) -> str:
            f = cooldown.model_dump().get(field)
            return f'cd.{f}'
Im trying to return a specific cd field which is a datetime thats why a projection to avoid getting the whole document

**Reply by mrrobinhood5** [reply]
Ok. heres the aggregation pipeline to accomplish what you need:
```py
field='rep'
[
{'$project':{'cd': f'$cooldown.{field}', '_id': 0}},
{'$replaceWith':{'$setField':{'field':f'cd.{field}', 'input':'$$ROOT', 'value':'$cd'}}},
{'$unset':'cd'}
]
```
Id say it cannot be done inside a beanie projection `BaseModel` because you don't have a way to pass in the `field` variable inside the `Settings` class. 
it should look something like this:
```py
cooldown = Cooldown()
await Player(player_id=1234, cooldown=cooldown).insert()
field = 'rep'
player = await Player.find(Player.player_id == 1234).aggregate(
    [
        {'$project':{'cd': f'$cooldown.{field}', '_id': 0}},
        {'$replaceWith':{'$setField':{'field':f'cd.{field}', 'input':'$$ROOT', 'value':'$cd'}}},
        {'$unset':'cd'}
    ]).to_list()
```
Alternatively you can use monggregate to build your pipeline dynamically somewhere else:
```py
from monggregate import Pipeline
a = Pipeline()
a.project(projection={'cd': f'$cooldown.{field}', '_id':0})
a.replace_with(document={'$setField':{'field':f'cd.{field}', 'input':'$$ROOT', 'value':'$cd'}})
a.unset(field='cd')
p = await Player.find(Player.player_id == 1234).aggregate(a.export()).to_list()
```
